runtime/runtime-params-estimator/src/vm_estimator.rs

```diff
@@ -130,6 +135,128 @@ pub fn load_and_compile(
 
 const USING_LIGHTBEAM: bool = cfg!(feature = "lightbeam");
 
+fn measure_contract(
+    vm_kind: VMKind,
+    gas_metric: GasMetric,
+    contract: &ContractCode,
+    cache: Option<&dyn CompiledContractCache>,
+) -> u64 {
+    let start = start_count(gas_metric);
+    let vm_config = VMConfig::default();
+    let result = precompile_contract_vm(vm_kind, &contract, &vm_config, cache);
+    assert!(result.is_ok(), "Compilation failed");
+    let end = end_count(gas_metric, &start);
+    end
+}
+
+/// Returns `(a, b)` - approximation coefficients for formula `a + b * x`
+/// where `x` is is the contract size in bytes. Practically, we compute upper bound
+/// of this approximation, assuming that all contract consists of code only.
+fn precompilation_cost(gas_metric: GasMetric, vm_kind: VMKind) -> (Ratio<i128>, Ratio<i128>) {
+    let workdir = tempfile::Builder::new().prefix("runtime_testbed").tempdir().unwrap();
+    let store = create_store(&get_store_path(workdir.path()));
+    let cache_store = Arc::new(StoreCompiledContractCache { store });
+    let cache: Option<&dyn CompiledContractCache> = Some(cache_store.as_ref());
+
+    let mut xs = vec![];
+    let mut ys = vec![];
+
+    // We use core-contracts, e2f60b5b0930a9df2c413e1460e179c65c8876e3.
+    // File 341191, code 279965, data 56627.
+    let data = include_bytes!("../test-contract/res/lockup_contract.wasm");
+    let contract = ContractCode::new(data.to_vec(), None);
+    xs.push(data.len() as u64);
+    ys.push(measure_contract(vm_kind, gas_metric, &contract, cache));
+
+    // File 257516, code 203545, data 50419.
+    let data = include_bytes!("../test-contract/res/staking_pool.wasm");
```

Comment:
Is it the right approach to use "real world" contracts for estimation? My gut feeling is that it probably is not. 

We should be estimating worst case, so we should prepare a synthetic contract that exercises this case. We *should* use real contracts to double-check that a synthetic worst-case is at least not better that real world code. 

However, just using real world contracts feels like we are saying "look, we don't know what the worst-case is, but we *assume* that it's within a stone's throw of typical case, so let's just measure that".

Reply:
I think max(m1,m2) approach will give you upper bound even for unknown code-portion contract, proof is:
```
let cost = max(m1,m2)*file_size+max(b1,b2)
then cost>=m1*filesize+b1>m1*codesize+b1=cost est based on codesize
cost>=m2*filesize+b2>m2*datasize=cost est based on datasize
```